man/tilestache-list.1
.TH TILESTACHE-LIST 1 "Nov 10, 2010"
.SH NAME
tilestache-list \- generates list of tiles based on geographic or other criteria
.SH SYNOPSIS
.B tilestache-list
.RI [ options ]
.RI [ zoom... ]
.SH DESCRIPTION
This manual page documents briefly the \fBtilestache-list\fR command.
.PP
\fBtilestache\-list\fP generates a list of tiles based on geographic or
other criteria.
.PP
No images are created and no Tilestache configuration is read, but a list
of tile coordinates in Z/X/Y form compatible with
\fBtilestache-seed \-\-tile-list\fP output.
.PP
Example:
.PP
    tilestache\-list.py \-b 52.55 13.28 52.46 13.51 11 12 13
.PP
Protip: seed a cache in parallel on 8 CPUs with split and xargs like this:
.PP
    tilestache\-list.py 12 13 14 15 | split \-l 20 \- tiles/list\-
    ls \-1 tiles/list\-* | xargs \-n1 \-P8 tilestache\-seed.py \-c tilestache.cfg \-l osm \-\-tile\-list
.SH OPTIONS
.TP
\fB-b\fR, \fB\-\-bbox\fR \fIlat lon lat lon\fR
Bounding box in floating point geographic coordinates:
south west north east.
.TP
\fB-p\fR, \fB\-\-padding\fR \fInumber\fR
Extra margin of tiles to add around bounded area.
.TP
\fB\-\-from\-mbtiles\fR \fIfile\fR
Optional input file for tiles, will be read as an MBTiles 1.1 tileset.
See http://mbtiles.org for more information.
Overrides \-\-bbox and \-\-padding.
.SH AUTHOR
\fBTileStache\fR was written by Michal Migurski <mike@stamen.com>.


CHANGELOG
2016-12-21: 1.51.4
- Support for GDAL 2.x types OFTInteger64 and OFTInteger64 on Vector Tiles

2016-10-30: 1.51.3
- Fixes for debian packaging

2016-10-27: 1.51.2
- Adding support for .pbf format for Mapbox Vector Tiles
- Configuration can be passed as a dictionary in addition to path to JSON file.

2016-09-01: 1.51.1
- Bugfix on Mapnik image object

2016-09-01: 1.51
- Replacing PIL by Pillow
- Added example with raster

2015-01-30: 1.50.1
- Fixed missing ttf file

2015-01-21: 1.50.0
- Added 'scale factor' option to Mapnik image provider
- Added S3 policy specification to config
- Added dictionary config to VecTiles config
- Added pixel effects to all layer configurations
- Added remote configuration to MapnikGrid to match Mapnik
- Fixed UtfGridComposite to work with latest TileStache
- Fixed Pillow dependency
- Fixed "cache lifespan" in Redis cache

2014-06-07: 1.49.11
- Fixed 'allow-origin header' getting overwritten (JesseCrocker)
- Added 'source projection' parameter to UrlTemplate provider (JesseCrocker)
- Fixed minor AttributeError when palette256 argument is None (PerryGeo)
- Added Timeout feature to Proxy and UrlTemplate providers (Juliomalegria)
- Adjusted curl examples (Peter Richardson)
- More specific error message when SpatialRef is missing (probabble)

2014-05-10: 1.49.10
- Fixed Travis build.
- Fixed import errors for case-insensitive filesystems.
- Added TileStache Vagrant machine configuration.
- Fixed some memcache testing problems.

2014-05-10: 1.49.9
- Moved everything to PyPI and fixed VERSION kerfuffle.

2013-07-02: 1.49.8
- Dropped Proxy provider srs=900913 check.
- Updated most JSON mime-types from text/json to application/json.
- Fixed requestHandler() query_string argument to match existing API docs.
- Switched order of simplify and transform in VecTiles queries to handle GEOS bug.
- Made Rummy CORS-compliant.

2013-06-23: 1.49.7
- Added VecTiles:MultiProvider to merge TopoJSON and GeoJSON output layers.

2013-06-21: 1.49.6
- Switched to ST_SimplifyPreserveTopology in VecTiles clipping to calm GEOS.

2013-06-20: 1.49.5
- Tweaked VecTiles Provider documentation.

2013-06-20: 1.49.4
- Added TopoJSON support to VecTiles Provider.

2013-06-15: 1.49.3
- Fixed runaway query bug in VecTiles Provider with looped bounds in query_columns().

2013-06-14: 1.49.2
- Fixed SQL bug in VecTiles Provider for limiting column queries.

2013-06-14: 1.49.1
- Updated magic column names in VecTiles Provider to __geometry__ and __id__ for feature ID.

2013-06-05: 1.49.0
- Added option to use S3 reduced redundancy storage for cheaper caching.
- Made Postgres connections per-request instead of long term in VecTiles Provider.

2013-05-28: 1.48.2
- Fixed a serious documentation incompatibility with TileStache.requestHandler() introduced in 1.47.0.

2013-05-27: 1.48.1
- Fixed Werkzeug dependency in setup.py.

2013-05-26: 1.48.0
- Fixed a serious documentation incompatibility with TileStache.getTile() introduced in 1.47.0.

2013-04-24: 1.47.4
- Fixed from-mbtiles/output-directory use case in tilestache-seed, which I had broken.

2013-04-22: 1.47.3
- Fixed clobbered content from TheTileLeftANote exception.

2013-04-22: 1.47.2
- Fixed getTile bugs introduced in 1.47.0.

2013-04-20: 1.47.1
- Fixed typo in setup.py.

2013-04-20: 1.47.0
- Exposed mutable HTTP status codes and headers to TileStache.getTile() and caches via TheTileLeftANote exception.
- Removed old layer.redirects implementation via TheTileIsInAnotherCastle exception.
- Improved PIL dependency in setup requirements to include Pillow, with patch from Fabian Büchler.
- Fixed coordinate generator usage in tilestache-seed with patch from Seth Fitzsimmons.
- Added Redis cache provider with patch from Matthew Perry.

2013-04-12: 1.46.10
- Prefixed most Postgis functions in VecTile provider with "ST_" for forward compatibility.

2013-04-12: 1.46.9
- Added ogr.OFTDate and ogr.OFTTime to list of acceptable field types in Vector provider.

2013-04-12: 1.46.8
- Added ogr.OFTDateTime to list of acceptable field types in Vector provider, with patch from probabble.

2013-03-27: 1.46.7
- Fixed a stupid error in VecTiles.server discarding important Postgres connection details.

2013-03-26: 1.46.6
- Improved error tolerance in VecTiles client and server.

2013-03-25: 1.46.5
- Added new simplify_until parameter to VecTiles.Provider.
- Added some invalid geometry checks to VecTiles.server.
- Added !bbox! token to VecTiles query parser.

2013-03-19: 1.46.4
- Added new zoom_data parameter to TileStache.Goodies.VecTiles Datasource.

2013-03-19: 1.46.3
- Added new simplify parameter to TileStache.Goodies.VecTiles Provider.

2013-03-14: 1.46.2
- Added new parameters to TileStache.Goodies.VecTiles Datasource and Provider.

2013-03-14: 1.46.1
- Fixed missing TileStache.Goodies.VecTiles by adding it to setup.py.

2013-03-12: 1.46.0
- Added TileStache.Goodies.VecTiles to support efficient vector data tiles for use in Mapnik.
- Fixed alpha-channel issue in Sandwich provider with patch from Lars Ahlzen.

2013-02-26: 1.45.0
- Made S3 credentials optional since Boto allows them in via environment vars or config file.
- Added path parameter to S3 cache, to use a subdirectory rather than bucket root.
- Eased bitmap size checks in Sandwich provider for retina use.
- Added "quadtile" directory layout for disk cache provider.
- Added JSON support to MBTiles provider from Aaron Cope.
- Added Travis CI tests for Vector layer.

2013-02-22: 1.44.0
- Merged Ragi Burhum's Travis continuous integration tests with ~1/3rd test coverage.
- Modified Proxy provider to return untouched original image if size and type match.
- Updated Mapnik usage to reflect recently released versions.
- Fixed Mapnik provider deadlock bug with patch from Lee Shepherd.
- Added "key prefix" argument to Memcache cache provider.

2013-01-03: 1.43.0
- Added new tilestache-list.py script to feed tilestache-seed.py --tile-list.

scripts/tilestache-list.py
#!/usr/bin/env python
"""tilestache-list.py will list your tiles.

This script is intended to be run directly. This example lists tiles in the area
around West Oakland (http://sta.mn/ck) in the "osm" layer, for zoom levels 12-15:

    tilestache-list.py -b 37.79 -122.35 37.83 -122.25 12 13 14 15

See `tilestache-list.py --help` for more information.
"""

from sys import stderr, path
from optparse import OptionParser

from TileStache.Core import KnownUnknown
from TileStache import MBTiles

from ModestMaps.Core import Coordinate
from ModestMaps.Geo import Location
from ModestMaps.OpenStreetMap import Provider

parser = OptionParser(usage="""%prog [options] [zoom...]

Generates a list of tiles based on geographic or other criteria. No images are
created and no Tilestache configuration is read, but a list of tile coordinates
in Z/X/Y form compatible with `tilestache-seed --tile-list` is output.

Example:

    tilestache-list.py -b 52.55 13.28 52.46 13.51 11 12 13

Protip: seed a cache in parallel on 8 CPUs with split and xargs like this:

    tilestache-list.py 12 13 14 15 | split -l 20 - tiles/list-
    ls -1 tiles/list-* | xargs -n1 -P8 tilestache-seed.py -c tilestache.cfg -l osm --tile-list

See `%prog --help` for info.""")

defaults = dict(padding=0, bbox=(37.777, -122.352, 37.839, -122.226))

parser.set_defaults(**defaults)

parser.add_option('-b', '--bbox', dest='bbox',
                  help='Bounding box in floating point geographic coordinates: south west north east. Default value is %.7f, %.7f, %.7f, %.7f.' % defaults['bbox'],
                  type='float', nargs=4)

parser.add_option('-p', '--padding', dest='padding',
                  help='Extra margin of tiles to add around bounded area. Default value is %s (no extra tiles).' % repr(defaults['padding']),
                  type='int')

parser.add_option('--from-mbtiles', dest='mbtiles_input',
                  help='Optional input file for tiles, will be read as an MBTiles 1.1 tileset. See http://mbtiles.org for more information. Overrides --bbox and --padding.')

def generateCoordinates(ul, lr, zooms, padding):
    """ Generate a stream of coordinates for seeding.
    
        Flood-fill coordinates based on two corners, a list of zooms and padding.
    """
    # start with a simple total of all the coordinates we will need.
    count = 0
    
    for zoom in zooms:
        ul_ = ul.zoomTo(zoom).container().left(padding).up(padding)
        lr_ = lr.zoomTo(zoom).container().right(padding).down(padding)
        
        rows = lr_.row + 1 - ul_.row
        cols = lr_.column + 1 - ul_.column
        
        count += int(rows * cols)

    # now generate the actual coordinates.
    # offset starts at zero
    offset = 0
    
    for zoom in zooms:
        ul_ = ul.zoomTo(zoom).container().left(padding).up(padding)
        lr_ = lr.zoomTo(zoom).container().right(padding).down(padding)

        for row in range(int(ul_.row), int(lr_.row + 1)):
            for column in range(int(ul_.column), int(lr_.column + 1)):
                coord = Coordinate(row, column, zoom)
                
                yield coord
                offset += 1

def tilesetCoordinates(filename):
    """ Generate a stream of (offset, count, coordinate) tuples for seeding.
    
        Read coordinates from an MBTiles tileset filename.
    """
    coords = MBTiles.list_tiles(filename)
    count = len(coords)
    
    for (offset, coord) in enumerate(coords):
        yield coord

if __name__ == '__main__':
    options, zooms = parser.parse_args()

    if bool(options.mbtiles_input):
        coordinates = MBTiles.list_tiles(options.mbtiles_input)

    else:
        lat1, lon1, lat2, lon2 = options.bbox
        south, west = min(lat1, lat2), min(lon1, lon2)
        north, east = max(lat1, lat2), max(lon1, lon2)

        northwest = Location(north, west)
        southeast = Location(south, east)
        
        osm = Provider()

        ul = osm.locationCoordinate(northwest)
        lr = osm.locationCoordinate(southeast)

        for (i, zoom) in enumerate(zooms):
            if not zoom.isdigit():
                raise KnownUnknown('"%s" is not a valid numeric zoom level.' % zoom)

            zooms[i] = int(zoom)
        
        if options.padding < 0:
            raise KnownUnknown('A negative padding will not work.')

        coordinates = generateCoordinates(ul, lr, zooms, options.padding)
    
    for coord in coordinates:
        print '%(zoom)d/%(column)d/%(row)d' % coord.__dict__
